sr_test_force_resolution/src/sr_test_force_resolution/sr_test_force_resolution.py

```diff
@@ -0,0 +1,451 @@
+# under the terms of the GNU General Public License as published by the Free
+# Software Foundation version 2 of the License.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+# FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
+# more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program. If not, see <http://www.gnu.org/licenses/>.
+
+import rospy
+import copy
+import termios
+import sys
+import csv
+import datetime
+from pynput import keyboard
+from control_msgs.msg import JointControllerState
+from sensor_msgs.msg import JointState
+from std_msgs.msg import Float64
+from sr_robot_commander.sr_hand_commander import SrHandCommander
+from sr_controllers_tools.sr_controller_helper import ControllerHelper
+from controller_manager_msgs.srv import (ListControllers, LoadController,
+                                         SwitchController, SwitchControllerRequest)
+
+
+class ControllerStateMonitor():
+    def __init__(self, name):
+        self.name = name
+        self.enable_output = False
+        self.subscriber = None
+        self.output_controller_state_keys = ['timestamp', 'command', 'setpoint', 'process_value', 'process_value_dot']
+        self.output_controller_state = {}
+        self.initialise_output_dictionary()
+
+    def initialise_output_dictionary(self):
+        for key in self.output_controller_state_keys:
+            self.output_controller_state[key] = []
+
+    def unsubscribe_all(self):
+        self.subscriber.unsubscribe()
+
+
+class TestForceResolution():
+    def __init__(self, keyboard_control=True, side="right"):
+        self.active_tests = []
+        self._controller_subscribers = {}
+        self._last_joint_state = JointState()
+        self._output_jointstate = {}
+        self._hand_prefix = side[0] + 'h_'
+        self._output_jointstate_keys = ['timestamp', 'position', 'velocity', 'effort']
+        self._fingers_with_j0 = ['ff', 'mf', 'rf', 'lf']
+        self._mode = ''
+        self._j0_position_controllers = ["sh_{0}{1}j0_position_controller".format(self._hand_prefix, joint)
+                                         for joint in self._fingers_with_j0]
+        self.requested_joints = []
+        self._joint_states_zero = {'rh_FFJ1': 0, 'rh_FFJ2': 0, 'rh_FFJ3': 0, 'rh_FFJ4': 0,
+                                   'rh_MFJ1': 0, 'rh_MFJ2': 0, 'rh_MFJ3': 0, 'rh_MFJ4': 0,
+                                   'rh_RFJ1': 0, 'rh_RFJ2': 0, 'rh_RFJ3': 0, 'rh_RFJ4': 0,
+                                   'rh_LFJ1': 0, 'rh_LFJ2': 0, 'rh_LFJ3': 0, 'rh_LFJ4': 0, 'rh_LFJ5': 0,
+                                   'rh_THJ1': 0, 'rh_THJ2': 0, 'rh_THJ3': 0, 'rh_THJ4': 0, 'rh_THJ5': 0,
+                                   'rh_WRJ1': 0, 'rh_WRJ2': 0}
+        self._joint_ranges = {'TH':
```

Comment:
can we get this from some parameters? it would be more flexible

Reply:
We could generate this dict automatically. The thinking here was that someone could easily change the joint limits for various tests